airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialReadUtil.java

```diff
@@ -0,0 +1,233 @@
+        }
+      });
+    }
+
+    final List<ConfiguredAirbyteStream> streamsForPkSync = new ArrayList<>();
+    fullCatalog.getStreams().stream()
+        .filter(stream -> streamsStillinPkSync.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream())))
+        .map(Jsons::clone)
+        .forEach(streamsForPkSync::add);
+    final List<ConfiguredAirbyteStream> newlyAddedStreams = identifyStreamsToSnapshot(fullCatalog, stateManager.getInitialStreamsSynced());
+    streamsForPkSync.addAll(newlyAddedStreams);
+
+    return new InitialLoadStreams(streamsForPkSync, pairToInitialLoadStatus);
+  }
+
+  private static List<ConfiguredAirbyteStream> identifyStreamsToSnapshot(final ConfiguredAirbyteCatalog catalog,
+      final Set<AirbyteStreamNameNamespacePair> alreadySyncedStreams) {
+    final Set<AirbyteStreamNameNamespacePair> allStreams = AirbyteStreamNameNamespacePair.fromConfiguredCatalog(catalog);
+    final Set<AirbyteStreamNameNamespacePair> newlyAddedStreams = new HashSet<>(Sets.difference(allStreams, alreadySyncedStreams));
+    return catalog.getStreams().stream()
+        .filter(c -> c.getSyncMode() == SyncMode.INCREMENTAL)
+        .filter(stream -> newlyAddedStreams.contains(AirbyteStreamNameNamespacePair.fromAirbyteStream(stream.getStream()))).map(Jsons::clone)
+        .collect(Collectors.toList());
+  }
+
+  // Build a map of stream <-> primary key info (primary key field name + datatype) for all streams currently undergoing initial primary key syncs.
+  private static Map<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair, PrimaryKeyInfo> initPairToPrimaryKeyInfoMap(
+      final InitialLoadStreams initialLoadStreams,
+      final Map<String, TableInfo<CommonField<MysqlType>>> tableNameToTable) {
+    final Map<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair, PrimaryKeyInfo> pairToPkInfoMap = new HashMap<>();
+    // For every stream that is in primary initial key sync, we want to maintain information about the current primary key info associated with the
+    // stream
+    initialLoadStreams.streamsForInitialLoad().forEach(stream -> {
+      final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair = new io.airbyte.protocol.models.AirbyteStreamNameNamespacePair(stream.getStream().getName(), stream.getStream().getNamespace());
+      final PrimaryKeyInfo pkInfo = getPrimaryKeyInfo(stream, tableNameToTable);
+      pairToPkInfoMap.put(pair, pkInfo);
+    });
+    return pairToPkInfoMap;
+  }
+
+  // Returns the primary key info associated with the stream.
+  private static PrimaryKeyInfo getPrimaryKeyInfo(final ConfiguredAirbyteStream stream, final Map<String, TableInfo<CommonField<MysqlType>>> tableNameToTable) {
+    // For cursor-based syncs, we cannot always assume a primary key field exists. We need to handle the case where it does not exist when we support
+    // cursor-based syncs.
+    final String pkFieldName = stream.getStream().getSourceDefinedPrimaryKey().get(0).get(0);
```

Comment:
Would we get to a point where this code path gets run when the source has no defined PK? I know in MySQL 8.0, it allows for invisible PK now.

Reply:
This case will need to be handled for cursor-based mode. We only allow incremental syncs in CDC mode if the table has a defined primary key, so we're safe here